### PR TITLE
[machine-learning-training] Add sample code for get layer and descrip…

### PR DESCRIPTION
**Pull request by songgot**
…tion for some layer type

- Add sample code for ml_train_model_get_layer()
- Add description for convld, lstmcell, grucell, rnncell and zoneout_lstemcell

Signed-off-by: Hyunil <hyunil46.park@samsung.com>

### CODE REVIEW

**Comment by songgot** [comment]
please review this patch. thank you.

**Review by arisha-srbd — COMMENTED** [review]
Hello @songgot. Hope you are doing well.
I see that you have made some additions to the table, so I wanted to know if you want me to review the whole PR or just check the updated parts. I shall do accordingly. Thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 1]
```diff
@@ -169,6 +184,18 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     |                             | softmax                     |                         | Softmax function
 &#xfeff;                                                     | return_sequences            | (boolean)                   | false                   | Return only the last output if true, else return full output
 &#xfeff;                                                     | dropout                     | (float)                     | 0                       | Dropout rate
+&#xfeff;                                                     | integrate_bias              | (boolean)                   | 0                       | Integrate bias_ih, bias_hh to bias_h
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:187]: okay, thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 2]
```diff
@@ -197,6 +241,41 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | reset_after                 | (boolean)                   | 1                       | Whether apply reset gate before/after the matrix
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:245]: okay, thank you.

**Comment by songgot** [comment]
@arisha-srbd , thank you.
Unfortunately, we do not have permission to this PR.
So Could you merge this PR after review on our side?
I'll let you know after the review is done.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 3]
```diff
@@ -147,6 +152,16 @@ Type | Key | Value | Default value | Description
...
 `ML_TRAIN_LAYER_TYPE_FC`                                     |                             |                             |                         | Fully connected layer
 &#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of outputs
+`ML_TRAIN_LAYER_TYPE_CONV1D` (since 7.0)                     |                             |                             |                         | 1D Convolution layer
+&#xfeff;                                                     | filters                     | (unsigned integer)          |                         | Number of filters
+&#xfeff;                                                     | kernel_size                 | (array of unsigned integer) |                         | Comma-seperated unsigned integers for kernel size, `height, width`  respectively
+&#xfeff;                                                     | stride                      | (array of unsigned integer) | 1, 1                    | Comma-seperated unsigned integers for strides, `height, width`  respectively
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:158]: same here for stride

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 4]
```diff
@@ -147,6 +152,16 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | padding                     | (categorical)               | valid                   | Padding type
+&#xfeff;                                                     |                             | valid                       |                         | No padding
+&#xfeff;                                                     |                             | same                        |                         | Preserve height/width dimension
+&#xfeff;                                                     |                             | (unsigned integer)          |                         | Size of padding applied uniformly to all side
+&#xfeff;                                                     |                             | (array of unsigned integer of size 2) |                         | Padding for height, width
+&#xfeff;                                                     |                             | (array of unsigned integer of size 4) |                         | Padding for top, bottom, left, right
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:164]: we do not need this.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 5]
```diff
@@ -147,6 +152,16 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | stride                      | (array of unsigned integer) | 1, 1                    | Comma-seperated unsigned integers for strides, `height, width`  respectively
+&#xfeff;                                                     | padding                     | (categorical)               | valid                   | Padding type
+&#xfeff;                                                     |                             | valid                       |                         | No padding
+&#xfeff;                                                     |                             | same                        |                         | Preserve height/width dimension
+&#xfeff;                                                     |                             | (unsigned integer)          |                         | Size of padding applied uniformly to all side
+&#xfeff;                                                     |                             | (array of unsigned integer of size 2) |                         | Padding for height, width
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:163]: It needs to be 
```Padding for left, right```

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 6]
```diff
@@ -169,6 +184,18 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     |                             | tanh                        |                         | Hyperbolic tangent
+&#xfeff;                                                     |                             | sigmoid                     |                         | Sigmoid function
+&#xfeff;                                                     |                             | relu                        |                         | Relu function
+&#xfeff;                                                     |                             | softmax                     |                         | Softmax function
+&#xfeff;                                                     | dropout                     | (float)                     | 0                       | Dropout rate
+&#xfeff;                                                     | integrate_bias              | (boolean)                   | false                   | Integrate bias_ih, bias_hh to bias_h
+&#xfeff;                                                     | max_timestep                | (unsigned integer)          | 0                       | Maximum timestep
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:197]: This needs to be removed.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 7]
```diff
@@ -169,6 +184,18 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | timestep                    | (unsigned integer)          |                         | Timestep for which lstm should operate
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:198]: This also needs to be removed. We do not need the time steps in Cell implementation.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 8]
```diff
@@ -182,7 +209,24 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     |                             | relu                        |                         | Relu function
 &#xfeff;                                                     |                             | softmax                     |                         | Softmax function
 &#xfeff;                                                     | return_sequences            | (boolean)                   | false                   | Return only the last output if true, else return full output
+&#xfeff;                                                     | bidirectional               | (boolean)                   | false                   | Option for bidirectional
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:212]: This should be removed. we are not going to release the bidirectional LSTM in M1.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 9]
```diff
@@ -500,6 +579,37 @@ status = ml_train_layer_set_property(layers[1], "name=outputlayer",
...
+}
+status = ml_train_layer_set_property(add_layer, "name=inputlayer", NULL);
+if (status != ML_ERROR_NONE) {
+  // handle error
+}
+status = ml_train_model_add_layer(model, add_layer);
+if (status != ML_ERROR_NONE) {
+  // handle error
+}
+status = ml_train_model_get_layer(model, "inputlayer", &get_layer);
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:610]: add
``` c++
if (status != ML_ERROR_NONE){
   // handle error
}
```

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 10]
```diff
@@ -500,6 +579,37 @@ status = ml_train_layer_set_property(layers[1], "name=outputlayer",
...
+  // handle error
+}
+status = ml_train_layer_set_property(add_layer, "name=inputlayer", NULL);
+if (status != ML_ERROR_NONE) {
+  // handle error
+}
+status = ml_train_model_add_layer(model, add_layer);
+if (status != ML_ERROR_NONE) {
+  // handle error
+}
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:609]: How about giving the notification saying
"layer object owned by model. Do not delete it"

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 11]
```diff
@@ -147,6 +152,16 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     | weight_regularizer_constant | (float)                     | 1                       | Weight regularizer constant
 `ML_TRAIN_LAYER_TYPE_FC`                                     |                             |                             |                         | Fully connected layer
 &#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of outputs
+`ML_TRAIN_LAYER_TYPE_CONV1D` (since 7.0)                     |                             |                             |                         | 1D Convolution layer
+&#xfeff;                                                     | filters                     | (unsigned integer)          |                         | Number of filters
+&#xfeff;                                                     | kernel_size                 | (array of unsigned integer) |                         | Comma-seperated unsigned integers for kernel size, `height, width`  respectively
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:157]: thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 5 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:163]: I will fix 'height, width ' to 'left, right'

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 4 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:164]: and I will remove this line.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 6 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:197]: okay, I will remove max_timestep

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 7 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:198]: okay

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 12]
```diff
@@ -197,6 +241,41 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     |                             | softmax                     |                         | Softmax function
+&#xfeff;                                                     | recurrent_activation        | (categorical)               | sigmoid                 | Activation type for recurrent step
+&#xfeff;                                                     |                             | tanh                        |                         | Hyperbolic tangent
+&#xfeff;                                                     |                             | sigmoid                     |                         | Sigmoid function
+&#xfeff;                                                     |                             | relu                        |                         | Relu function
+&#xfeff;                                                     |                             | softmax                     |                         | Softmax function
+&#xfeff;                                                     | cell_state_zoneout_rate     | (float)                     | 0                       | Zoneout rate for cell state
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:273]: thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 10 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:609]: okay

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 9 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:610]: okay

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lhs8928) [reply docs/application/native/guides/machine-learning/machine-learning-training.md]: Can we change the description of reset_after to "apply reset gate before/after the matrix multiplication"?

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 13]
```diff
@@ -147,6 +152,15 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     | weight_regularizer_constant | (float)                     | 1                       | Weight regularizer constant
 `ML_TRAIN_LAYER_TYPE_FC`                                     |                             |                             |                         | Fully connected layer
 &#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of outputs
+`ML_TRAIN_LAYER_TYPE_CONV1D` (since 7.0)                     |                             |                             |                         | 1D Convolution layer
+&#xfeff;                                                     | filters                     | (unsigned integer)          |                         | Number of filters
+&#xfeff;                                                     | kernel_size                 | (unsigned integer)          |                         | Comma-seperated unsigned integers for kernel size, `height, width`  respectively
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:157]: should remove `height, width`. It is single unsigned integer.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 14]
```diff
@@ -147,6 +152,15 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | stride                      | (unsigned integer)          |                         | Comma-seperated unsigned integers for strides, `height, width`  respectively
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:158]: same here.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 15]
```diff
@@ -147,6 +152,15 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     | padding                     | (categorical)               | valid                   | Padding type
+&#xfeff;                                                     |                             | valid                       |                         | No padding
+&#xfeff;                                                     |                             | same                        |                         | Preserve height/width dimension
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:161]: Should remove height/width.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 13 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:157]: thank you

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 14 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:158]: thank you

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 15 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:161]: okay, thank you

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 16]
```diff
@@ -197,6 +237,42 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     |                             | softmax                     |                         | Softmax function
 &#xfeff;                                                     | return_sequences            | (boolean)                   | false                   | Return only the last output if true, else return full output
 &#xfeff;                                                     | dropout                     | (float)                     | 0                       | Dropout rate
+&#xfeff;                                                     | integrate_bias              | (boolean)                   | false                   | Integrate bias_ih, bias_hh to bias_h
+&#xfeff;                                                     | reset_after                 | (boolean)                   | true                    | Whether apply reset gate before/after the matrix
```
Comment (lhs8928) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:241]: I guess it's better to remove "Whether"

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 17]
```diff
@@ -197,6 +237,42 @@ Type | Key | Value | Default value | Description
...
+`ML_TRAIN_LAYER_TYPE_GRUCELL` (since 7.0)                    |                             |                             |                         | GRUCELL layer
+&#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of output neurons
+&#xfeff;                                                     | reset_after                 | (boolean)                   | true                    | Whether apply reset gate before/after the matrix multiplication
```
Comment (lhs8928) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:244]: same here

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 18]
```diff
@@ -197,6 +237,42 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     |                             | softmax                     |                         | Softmax function
+&#xfeff;                                                     | cell_state_zoneout_rate     | (float)                     | 0                       | Zoneout rate for cell state
+&#xfeff;                                                     | hidden_state_zoneout_rate   | (float)                     | 0                       | zoneout rate for hidden state
+&#xfeff;                                                     | integrate_bias              | (boolean)                   | false                   | Integrate bias_ih, bias_hh to bias_h
+&#xfeff;                                                     | test                        | (boolean)                   | false                   | Test mode
+&#xfeff;                                                     | max_timestep                | (unsigned integer)          |                         | Maximum timestep
```
Comment (lhs8928) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:273]: max_timestep, timestep property should be removed from zoneout lstmcell

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 19]
```diff
@@ -197,6 +237,42 @@ Type | Key | Value | Default value | Description
...
+&#xfeff;                                                     |                             | relu                        |                         | Relu function
+&#xfeff;                                                     |                             | softmax                     |                         | Softmax function
+&#xfeff;                                                     | cell_state_zoneout_rate     | (float)                     | 0                       | Zoneout rate for cell state
+&#xfeff;                                                     | hidden_state_zoneout_rate   | (float)                     | 0                       | zoneout rate for hidden state
+&#xfeff;                                                     | integrate_bias              | (boolean)                   | false                   | Integrate bias_ih, bias_hh to bias_h
+&#xfeff;                                                     | test                        | (boolean)                   | false                   | Test mode
```
Comment (lhs8928) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:272]: test property is not for user. please remove this property

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 18 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:273]: okay, thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 19 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:272]: okay, thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 17 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:244]: okay, thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 16 above)
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:241]: okay, thank you.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 20]
```diff
@@ -147,6 +152,15 @@ Type | Key | Value | Default value | Description
...
 `ML_TRAIN_LAYER_TYPE_FC`                                     |                             |                             |                         | Fully connected layer
 &#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of outputs
+`ML_TRAIN_LAYER_TYPE_CONV1D` (since 7.0)                     |                             |                             |                         | 1D Convolution layer
+&#xfeff;                                                     | filters                     | (unsigned integer)          |                         | Number of filters
+&#xfeff;                                                     | kernel_size                 | (unsigned integer)          |                         | Comma-seperated unsigned integers for kernel size
+&#xfeff;                                                     | stride                      | (unsigned integer)          |                         | Comma-seperated unsigned integers for strides
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:158]: okay, thank you.
Kernel size and Strides look good.

[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:158]: okay, thank you.
Kernel size and Strides look good.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 21]
```diff
@@ -147,6 +152,15 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     | weight_regularizer_constant | (float)                     | 1                       | Weight regularizer constant
 `ML_TRAIN_LAYER_TYPE_FC`                                     |                             |                             |                         | Fully connected layer
 &#xfeff;                                                     | unit                        | (unsigned integer)          |                         | Number of outputs
+`ML_TRAIN_LAYER_TYPE_CONV1D` (since 7.0)                     |                             |                             |                         | 1D Convolution layer
+&#xfeff;                                                     | filters                     | (unsigned integer)          |                         | Number of filters
+&#xfeff;                                                     | kernel_size                 | (unsigned integer)          |                         | Comma-seperated unsigned integers for kernel size
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:157]: okay, thank you.

**Review by lhs8928 — APPROVED** [review]
LGTM

**Review by arisha-srbd — COMMENTED** [review]
@songgot hello. Hope you are doing okay. This PR seems approved by @lhs8928. Do you want us to go ahead and merge the PR?

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Comment by songgot** [comment]
hello @arisha-srbd , Could you merge this PR?

**Review by arisha-srbd — APPROVED** [review]
@songgot hello. The PR will be merged today. 

@asm-foysal @mijin85cho I am approving this PR for merging.

**Review by arisha-srbd — COMMENTED** [review]
@songgot hello. Hope you are doing well! 
It seems like this PR contains remarks about Tizen 7.0. Do you think it'd be ok to publish information regarding 7.0 before its release? I should let you know that once this PR will be merged to the master (live) branch, the information would be published in docs.tizen.org. Let me know what you think.

**Pull request merged**